core/lib/mini_merkle_tree/src/lib.rs

```diff
@@ -86,67 +120,146 @@ where
+    /// `index` is relative to the leftmost uncached leaf.
+    /// # Panics
+    /// Panics if `index` is >= than the number of leaves in the tree.
+    pub fn merkle_root_and_path(&self, index: usize) -> (H256, Vec<H256>) {
+        assert!(index < self.hashes.len(), "leaf index out of bounds");
+        let mut end_path = vec![];
+        let root_hash = self.compute_merkle_root_and_path(index, Some(&mut end_path), None);
+        (
+            root_hash,
+            end_path.into_iter().map(Option::unwrap).collect(),
+        )
+    }
+
+    /// Returns the root hash and the Merkle proofs for a range of leafs.
+    /// The range is 0..length, where `0` is the leftmost untrimmed leaf.
+    /// # Panics
+    /// Panics if `length` is 0 or greater than the number of leaves in the tree.
+    pub fn merkle_root_and_paths_for_range(
+        &self,
+        length: usize,
+    ) -> (H256, Vec<Option<H256>>, Vec<Option<H256>>) {
+        assert!(length > 0, "range must not be empty");
+        assert!(length <= self.hashes.len(), "not enough leaves in the tree");
+        let mut right_path = vec![];
+        let root_hash =
+            self.compute_merkle_root_and_path(length - 1, Some(&mut right_path), Some(Side::Right));
+        (root_hash, self.cache.clone(), right_path)
+    }
+
+    /// Adds a raw hash to the tree (replaces leftmost empty leaf).
+    /// If the tree is full, its size is doubled.
+    /// Note: empty leaves != zero leaves.
+    pub fn push_hash(&mut self, leaf_hash: H256) {
+        self.hashes.push_back(leaf_hash);
+        if self.start_index + self.hashes.len() > self.binary_tree_size {
+            self.binary_tree_size *= 2;
+            if self.cache.len() < tree_depth_by_size(self.binary_tree_size) {
+                self.cache.push(None);
+            }
+        }
+    }
+
+    /// Adds a new leaf to the tree (replaces leftmost empty leaf).
+    /// If the tree is full, its size is doubled.
+    /// Note: empty leaves != zero leaves.
+    pub fn push(&mut self, leaf: [u8; LEAF_SIZE]) {
+        let leaf_hash = self.hasher.hash_bytes(&leaf);
+        self.push_hash(leaf_hash);
+    }
+
+    /// Caches the rightmost `count` leaves.
+    /// Does not affect the root hash, but makes it impossible to get the paths to the cached leaves.
+    /// # Panics
+    /// Panics if `count` is greater than the number of non-cached leaves in the tree.
+    pub fn trim_start(&mut self, count: usize) {
+        assert!(self.hashes.len() >= count, "not enough leaves to cache");
+        let mut new_cache = vec![];
+        // Cache is a left subset of the path to the first untrimmed leaf.
+        let root = self.compute_merkle_root_and_path(count, Some(&mut new_cache), Some(Side::Left));
+        self.hashes.drain(..count);
+        self.start_index += count;
+        if self.start_index == self.binary_tree_size {
+            // In order to be able to get the root of a completely trimmed tree,
+            // we need to cache it.
```

Comment:
Bikeshedding: IMO, this comment is confusing. The root is added not in any case when the tree is completely trimmed, but if it is completely trimmed *and* pushing a next element will lead to tree depth growth (since `binary_tree_size` is always a power of 2). IMO, it could be useful to document an invariant what `cache` is, at the corresponding field (instead of in `trim_start()`), so that it's more clear:

> Left subset of the Merkle path to the first untrimmed leaf (i.e., a leaf with index `self.start_index`). Used to fill in data for trimmed tree leaves when computing Merkle paths and the root hash. Because only the left subset of the path is used, the cache is not invalidated when new leaves are pushed into the tree. If all leaves are trimmed, `cache` is the left subset of the Merkle path to the next leaf to be inserted, which still has index `self.start_index`.